src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1962,6 +1963,151 @@ static TypeDeclarationSyntax constructTypeDeclaration(ContextAwareSyntax syntaxF
+                    constraints = _pool.Allocate<TypeParameterConstraintClauseSyntax>();
+                    this.ParseTypeParameterConstraintClauses(constraints);
+                }
+
+                _termState = outerSaveTerm;
+
+                SyntaxToken? openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
+                SyntaxToken? closeBrace;
+
+                // ignore members if missing open curly
+                if (openBrace.IsMissing)
+                {
+                    parseMembers = false;
+                }
+
+                if (parseMembers)
+                {
+                    members = _pool.Allocate<MemberDeclarationSyntax>();
+
+                    while (true)
+                    {
+                        SyntaxKind kind = this.CurrentToken.Kind;
+
+                        if (CanStartMember(kind))
+                        {
+                            // This token can start a member -- go parse it
+                            var saveTerm = _termState;
+                            _termState |= TerminatorState.IsPossibleMemberStartOrStop; // TODO2 add affected test
+
+                            MemberDeclarationSyntax member = this.ParseMemberDeclaration(SyntaxKind.ExtensionContainer);
+                            if (member != null)
+                            {
+                                members.Add(member);
+                            }
+                            else
+                            {
+                                this.SkipBadMemberListTokens(ref openBrace, members);
+                            }
+
+                            _termState = saveTerm;
+                        }
+                        else if (kind is SyntaxKind.CloseBraceToken or SyntaxKind.EndOfFileToken || this.IsTerminator())
+                        {
+                            // This marks the end of members of this extension container
+                            break;
+                        }
+                        else
+                        {
+                            // Error -- try to sync up with intended reality
+                            this.SkipBadMemberListTokens(ref openBrace, members);
+                        }
+                    }
+                }
+
+                if (openBrace.IsMissing)
+                {
+                    closeBrace = SyntaxFactory.MissingToken(SyntaxKind.CloseBraceToken);
+                    closeBrace = WithAdditionalDiagnostics(closeBrace, this.GetExpectedTokenError(SyntaxKind.CloseBraceToken, this.CurrentToken.Kind));
+                }
+                else
+                {
+                    closeBrace = this.EatToken(SyntaxKind.CloseBraceToken);
+                }
+
+                var modifiersList = (SyntaxList<SyntaxToken>)modifiers.ToList();
+                var membersList = (SyntaxList<MemberDeclarationSyntax>)members;
+                var constraintsList = (SyntaxList<TypeParameterConstraintClauseSyntax>)constraints;
```

Comment:
this is confusing.  we have the .ToListAndFree helpers.  Why not use those?  then you don't need a try/finally either.